woocommerce/woocommerce
# Sync the product attributes lookup table on product creation/change/deletion

### All Submissions:


### Changes proposed in this Pull Request:

Syncrhonizes the product attributes lookup table whenever a product or variation is created, modified or deleted. Syncrhonization happens whenever the lookup table exists, even if the usage of the table isn't active. The filtering is showcased in #29190, the initial table creation and filling is implemented in #29778, and the actual usage of the table for filtering is implemented in #29896.

Entries in the lookup table are created when:

* A new product is created
* A new variation is added to a product
* A product or variation that had been trashed is untrashed
* The product visibility changes to "search" or "hidden"
* The attributes for a product change, this includes:
  * New attributes are added or deleted for the product
  * Terms are added or removed from the attribute definition for the product
  * The attribute changes between "used for variations" and not
* The attributes for a variation change

Entries in the lookup table are deleted when:

* A product  or variation is trashed or permanently deleted
* The product visibility changes to "visible" or "catalog"

Entries in the lookup table are updated only when the stock status of a product or variation changes.

#### Scheduled vs direct updates

By default changes to the lookup table are not performed directly when the product changes, but a scheduled action is used instead. The name of the hook is `woocommerce_run_product_attribute_lookup_update_callback` (the `DataRegenerator` class used a hook with the same name, which has been renamed to `woocommerce_run_product_attribute_lookup_regeneration_callback`).

WooCommerce is prone to save stuff multiple times in a row, therefore a check is done before scheduling the action to prevent duplicates (existing pending action for same product and type of action, create/update/delete).

A new option is introduced, `woocommerce_attribute_lookup__direct_updates`. When set to `yes` updates are done directly when the change happens, skipping action scheduler; this may be useful for testing.

#### Per product regeneration

The "Regenerate the product attributes lookup table" tool in the Status - Tools page has been extended to include a product selector. If a product is selected then the "Regenerate" button will regenerate the data only for the product selected (directly, without scheduled action); if the selector is left blank then the table is fully regenerated using scheduled actions, as it was doing before.

![image](https://user-images.githubusercontent.com/937723/125607277-b5a9d837-0779-423c-b276-0e1429518f2a.png)

The single product regeneration may be useful if for some reason the lookup data ends up being incorrect for some reason, and ideally it would never be needed.

#### Changes in settings UI

When the feature is enabled (via `wc_get_container()->get(\Automattic\WooCommerce\Internal\ProductAttributesLookup\LookupDataStore::class)->show_feature();`) and the lookup table exists a new "Advanced" section is added to the product settings:

![image](https://user-images.githubusercontent.com/937723/124109557-90c13580-da67-11eb-8d17-50fd348455eb.png)

If a data regeneration is in progress then the section is still displayed but with no actionable options:

![image](https://user-images.githubusercontent.com/937723/124109913-f0b7dc00-da67-11eb-93f9-f3947a98d044.png)

Closes #29609.

### How to test the changes in this Pull Request:

1. Create and fill the lookup table according to the instructions of #29190.
2. Create/delete/modify products and variations according to the cases listed above, and verify that the lookup table updates accordingly.
3. Delete the lookup table (from the tools page) and verify that everything continues working.
4. To test the single product regeneration you can manually delete the lookup data for the product before using the regeneration tool and then check that the data was regenerated.

### Other information:


### Changelog entry

> Add - Product attributes lookup table synchronization when the table exists.
